src/cascadia/TerminalApp/TerminalTab.cpp

```diff
@@ -907,64 +907,77 @@ namespace winrt::TerminalApp::implementation
     {
-        auto weakThis{ get_weak() };
-        auto dispatcher = TabViewItem().Dispatcher();
         ControlEventTokens events{};
 
-        events.titleToken = control.TitleChanged([dispatcher, weakThis](auto&&, auto&&) -> winrt::fire_and_forget {
-            co_await wil::resume_foreground(dispatcher);
-            // Check if Tab's lifetime has expired
-            if (auto tab{ weakThis.get() })
-            {
-                // The title of the control changed, but not necessarily the title of the tab.
-                // Set the tab's text to the active panes' text.
-                tab->UpdateTitle();
-            }
-        });
-
-        events.colorToken = control.TabColorChanged([dispatcher, weakThis](auto&&, auto&&) -> winrt::fire_and_forget {
-            co_await wil::resume_foreground(dispatcher);
-            if (auto tab{ weakThis.get() })
-            {
-                // The control's tabColor changed, but it is not necessarily the
-                // active control in this tab. We'll just recalculate the
-                // current color anyways.
-                tab->_RecalculateAndApplyTabColor();
-            }
-        });
-
-        events.taskbarToken = control.SetTaskbarProgress([dispatcher, weakThis](auto&&, auto&&) -> winrt::fire_and_forget {
-            co_await wil::resume_foreground(dispatcher);
-            // Check if Tab's lifetime has expired
-            if (auto tab{ weakThis.get() })
-            {
-                tab->_UpdateProgressState();
-            }
-        });
+        events.titleToken = control.TitleChanged({ get_weak(), &TerminalTab::_controlTitleChanged });
+        events.colorToken = control.TabColorChanged({ get_weak(), &TerminalTab::_controlTabColorChanged });
+        events.taskbarToken = control.SetTaskbarProgress({ get_weak(), &TerminalTab::_controlSetTaskbarProgress });
+        events.readOnlyToken = control.ReadOnlyChanged({ get_weak(), &TerminalTab::_controlReadOnlyChanged });
+        events.focusToken = control.FocusFollowMouseRequested({ get_weak(), &TerminalTab::_controlFocusFollowMouseRequested });
 
-        events.readOnlyToken = control.ReadOnlyChanged([dispatcher, weakThis](auto&&, auto&&) -> winrt::fire_and_forget {
-            co_await wil::resume_foreground(dispatcher);
-            if (auto tab{ weakThis.get() })
-            {
-                tab->_RecalculateAndApplyReadOnly();
-            }
-        });
+        _controlEvents[paneId] = events;
+    }
 
-        events.focusToken = control.FocusFollowMouseRequested([dispatcher, weakThis](auto&& sender, auto&&) -> winrt::fire_and_forget {
-            co_await wil::resume_foreground(dispatcher);
-            if (const auto tab{ weakThis.get() })
+    winrt::fire_and_forget TerminalTab::_controlTitleChanged(Windows::Foundation::IInspectable sender, Microsoft::Terminal::Control::TitleChangedEventArgs e)
+    {
+        auto weakThis{ get_weak() };

```

Comment:
@lhecker you have any clever ideas on how to like, do this with some sort of smart template? Like, I want to just say 

```
events.titleToken = control.TitleChanged(til::OnForeground{ get_weak(), TabViewItem().Dispatcher(), &TerminalTab::_controlTitleChanged });
```

But methinks that wouldn't work, cause I'd have to keep the `til::OnForeground` somewhere to keep the inner function pointer alive.

Reply:
A `til::OnForeground` struct would look very similar btw. Like this basically (pseudocode):
```cpp
template<typename Weak, typename Func>
struct OnForeground {
    OnForeground(Weak weak, CoreDispatcher dispatcher, Func func) :
        _weak{ std::move(weak) },
        _dispatcher{ std::move(dispatcher) },
        _func{ std::move(func) }
    {}

    void operator()(auto sender, auto&&) const -> winrt::fire_and_forget
    {
        co_await wil::resume_foreground(dispatcher);

        if (const auto strong = _weak.get())
        {
            if constexpr (std::is_invocable_v<decltype(func), decltype(strong), decltype(sender)>)
            {
                (strong.get()->*_func)(sender);
            }
            else
            {
                (strong.get()->*_func)();
            }
        }
    }

private:
    Weak _weak;
    CoreDispatcher _dispatcher;
    Func _func;
};
```

This is basically identical to what the compiler generators for the "generator" lambdas. It has the same downsides too. It's a bit like a `std::mem_fn`, whose use is equally discouraged, because all these wrappers use dynamic dispatch for `func`, whereas the lambdas use static dispatch.